merge.py

```diff
@@ -1,29 +1,81 @@
-from typing import Callable
+from functools import partial
+from typing import Any, Callable, Optional, Sequence, Tuple
 
 
-def merge(dict1: dict, dict2: dict, array_strategy: Callable = lambda x, y: x) -> dict:
-    output = {}
+def merge(
+    dict1: dict, 
+    dict2: dict, 
+    dict_strategy: Callable = lambda x, y: x,
+    list_strategy: Callable = lambda x, y: x
+    ) -> dict:
+    """
+    `merge` is a function to recursively, uhh, merge two dictionaries
 
-    # adds keys from `dict2` if they do not exist in `dict1`
-    intersection = {**dict2, **dict1}
+    By default, `merge` operates like the expression ``{**dict1, **dict2}``, 
+    but when supplied functions for `dict_strategy` and `list_strategy` it 
+    is capable of merging any Python dict in any way imaginable. You can 
+    choose from some predefined functions from the `strategies` module to 
+    get started quickly, or roll your own for even more fun. 
 
-    for k_intersect, v_intersect in intersection.items():
-        if k_intersect not in dict1:
-            v_dict2 = dict2[k_intersect]
-            output[k_intersect] = v_dict2
+    TODO: add `strategies` module
+    """
 
-        elif k_intersect not in dict2:
-            output[k_intersect] = v_intersect
+    keys = get_keys(dict1, dict2)
+    return {
+        key: merge_map(
+            *get_values(dict1, dict2, key), 
+            dict_strategy, 
+            list_strategy)
+        for key in keys
```

Comment:
How does this work, you iterate over every key and create a merge_map output for that key? Why is the for loop under the function?

Reply:
```python
keys = get_keys(dict1, dict2)

return {
    key: merge_map(
        *get_values(dict1, dict2, key), 
        dict_strategy, 
        list_strategy)
    for key in keys
}
```
What this is doing is iterating over the union/superset of keys between the two dicts, and for each key it's creating:
- A key defined by `key` (nothing special here)
- The `key`'s value defined by calling `merge_map`
`merge_map` essentially just says, given a `key`, what's the corresponding value from `dict1` and `dict2` and which value should be returned to conform with our definition of a merged dict? There's some extra complexity due to the `dict_strategy` and `array_strategy` higher-order parameters, but if we just look at the one for dicts, it just says "if the key's corresponding value is a dict, we need to start the `merge` process all over again on it".
And the loop is at the end because this is a dict comprehension (documentation buried a bit https://docs.python.org/3.8/tutorial/datastructures.html#dictionaries)